### ISSUE TITLE
Handle & unit test all custom parameters

### ISSUE THREAD

**Issue by belluzj** [googlei18n/glyphsLib#253]
There are four lists of parameters to go through:
* those supported by Glyphs (e.g. listed in the Glyphs handbook): https://glyphsapp.com/downloads/handbook/Glyphs-Handbook-2.3.pdf
* those defined in UFO: http://unifiedfontobject.org/versions/ufo3/
* those that are understood by `ufo2ft`: https://github.com/googlei18n/ufo2ft
* those that are understood by `fontmake`: https://github.com/googlei18n/fontmake

For UFO and Glyphs parameters, we should have unit tests for each of them and decide where/under which key it is stored and how its value is converted when going from one format to the other.

Whenever the UFO standard doesn't support some Glyphs data, we should preferably translate it into a key/a format that `ufo2ft` or `fontmake` can process.

@moyogo

**Comment by moyogo** [comment]
@schriftgestalt Can you share the list of customParameters you mentioned in https://forum.glyphsapp.com/t/old-style-kern-table-export/3322/9 ?

Which customParameters are meaningful to have multiple times?

Here are the customParameters I was able to add through the list in the UI in 2.5b (1077):

Custom Parameter | Object(s) | Type(s)
-- | -- | --
ascender | master | unicode
Autohint | instance | int
Axes | font | list
Axis Location | master | list
blueScale | font | float
blueShift | font | int
capHeight | master | unicode
CJK Grid | master | int
CJK Grid Horizontal | master | int
CJK Grid Vertical | master | int
CJK Guide | master | unicode
codePageRanges | font, instance | list
Color Layers to SVG | instance | int
Color Palette for SVG | instance | int
Color Palettes | master | list
Compatible Name Table | font, instance | bool, int
compatibleFullName | instance | unicode
copyright | instance | unicode
CVT Table | master | unicode
Decompose Glyphs | instance | list
descender | master | unicode
description | font | unicode
Disable autohinting for glyphs | font, instance | list
Disable Last Change | font | bool
Disable Masters | instance | unicode
Disable Subroutines | font, instance | bool, int
Don't use Production Names | font, instance | bool, int
EditView Line Height | font | float
Export COLR Table | instance | int
Export Glyphs | instance | list
Export sbix Table | instance | int
Export STAT Table | instance | int
Export SVG Table | instance | int
Family Alignment Zones | font | list
familyName | instance | unicode
fileName | instance | unicode
Filter | instance | unicode
Fpgm Table Assembly | master | unicode
fsType | font, instance | list
GASP Table | font, instance | dict
Get Hints From Master | font | unicode
glyphOrder | font | list
Grid Spacing | instance | int
Hangul Composition Groups | font | dict
Has WWS Names | font, instance | bool, int
hheaAscender | master, instance | int
hheaDescender | master, instance | int
hheaLineGap | master, instance | int
Instance Preview | font, instance | list
InterpolationWeightY | instance | unicode
isFixedPitch | font | bool
italicAngle | master, instance | unicode
Keep Glyphs | instance | list
Keep Overlapping Components | instance | int
Keep Transformed Components | instance | int
license | font, instance | unicode
licenseURL | font, instance | unicode
Link Metrics With First Master | master | int
Link Metrics With Master | master | unicode
Local Interpolation | instance | unicode
localizedDesigner | font, instance | unicode
localizedFamilyName | font, instance | unicode
localizedStyleMapFamilyName | instance | unicode
localizedStyleName | instance | unicode
Make morx table | font | bool
makeOTF Argument | font, instance | unicode
Master Background Color | master | unicode
Master Color | master | unicode
Master Name | master | unicode
Master Stroke Color | master | unicode
Name Table Entry | font, instance | unicode
note | font | unicode
openTypeNameDescription | font | unicode
Optical Size | instance | unicode
panose | font, instance | list
Post Table Type | instance | int
postscriptFontName | instance | unicode
postscriptFullName | instance | unicode
preferredFamilyName | instance | unicode
preferredSubfamilyName | instance | unicode
PreFilter | instance | unicode
Prep Table Assembly | master | unicode
Preview Ascender | master | int
Preview Descender | master | int
Reencode Glyphs | instance | list
Remove Classes | instance | list
Remove Features | instance | list
Remove Glyphs | instance | list
Remove post names for webfonts | instance | int
Remove Prefixes | instance | unicode
Rename Glyphs | instance | list
Replace Class | instance | unicode
Replace Feature | instance | unicode
Replace Prefix | instance | unicode
ROS | font | unicode
sampleText | font, instance | unicode
Save as TrueType | instance | int
Scale to UPM | instance | unicode
shoulderHeight | master | int
smallCapHeight | master | int
styleMapFamilyName | instance | unicode
trademark | font, instance | unicode
TrueType Curve Error | font, instance | float, int
TTFAutohint binary path | font | unicode
TTFAutohint control instructions | instance | unicode
TTFAutohint options | instance | unicode
TTFOvershootSuppressionBelowPPM | font | int
TTFStems | master | list
TTFStemSetting | instance | list
TTFZones | master | list
typoAscender | master, instance | int
typoDescender | master, instance | int
typoLineGap | master, instance | int
underlinePosition | master, instance | int
underlineThickness | master, instance | int
unicodeRanges | font, instance | list
uniqueID | font, instance | unicode
unitsPerEm | instance | int
Update Features | instance | int
Use Extension Kerning | font | bool
Use Line Breaks | font | bool
Use Typo Metrics | font, instance | bool, int
Variation Font Origin | font | unicode
vendorID | font | unicode
versionString | font, instance | unicode
vheaVertAscender | master, instance | int
vheaVertDescender | master, instance | int
vheaVertLineGap | master, instance | int
Virtual Master | font | list
Webfont Formats | instance | unicode
Webfont Only | instance | int
weightClass | instance | int
winAscent | master, instance | int
winDescent | master, instance | int
WWSFamilyName | instance | unicode
WWSSubfamilyName | instance | unicode
xHeight | master | unicode

The following customParameters are mentioned in the [Handbook](https://glyphsapp.com/downloads/handbook/Glyphs-Handbook-2.3.pdf#page=180) but not listed in the UI:

Custom Parameter | Object(s) | Type(s)
-- | -- | --
DisableAllAutomaticBehaviour | font? | bool
Export Path | instance? | unicode
widthClass | instance? | unicode?
Write Kern Table | instance? | bool

**Comment by belluzj** [comment]
Note: have a look there: Glyphs.app/Contents/Frameworks/GlyphsCore.framework/Versions/A/

**Comment by schriftgestalt** [comment]
Yes. There are four files in `Glyphs.app/Contents/Frameworks/GlyphsCore.framework/Versions/A/Resources/`

- customParameterTypes.plist: contains all available keys with there type
- customFontParameters.plist: all keys that are relevant in the font
- customMasterParameters.plist: for masters
and you guessed it:
- customInstanceParameters.plist for the instances.

**Comment by moyogo** [comment]
@schriftgestalt Can we reuse the files here? Would you mind doing something like GlyphsData.xml that can be publicly shared?
